Тебайкин Максим

gODeaLoAple opened the pull request:
@hariton27sy

Commit:
Init commit

Commit:
Add tests and realisation

Commit:
Add images generation and readme.md

Commit:
Refactor color generators

Commit:
Refactor drawer so now it only draw rectangles, but not generate them

Commit:
Refactor tests with introduce _rectangles variable that contains all created rectangles per test

Commit:
Fix drawer when rectangles length is zero

Commit:
Add test fails logger and drawer

Commit:
Add speed check tests

Commit:
Tiny refactoring

Commit:
Refactor TagsCloudDrawer interface so that now it only draw rectangles on bitmap. Add disposing bitmap and graphics constructions

hariton27sy commented on cs/TagsCloudVisualization/SquareCalculator.cs:
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Функции используются как вспомогательные для тестов. Возможно стоит перенести в проект с тестами в таком случае. Иначе непонятно, есть проект с генератором облаков, в котором есть какие-то левые функции

gODeaLoAple replied:
А есть ли смысл перенести это не в Tests, а в отдельный проект, скажем, Utils?

hariton27sy commented on cs/TagsCloudVisualization/DistantPointFinder.cs:
```diff
@@ -0,0 +1,23 @@
+﻿using System.Collections.Generic;
```
Аналогично SquareCalculator. Функции используются как вспомогательные для тестов

hariton27sy commented on cs/TagsCloudVisualization/CircularVectorsGenerator.cs:
```diff
@@ -0,0 +1,34 @@
+            _lastAngle = 0;
+        }
+
+        public IEnumerable<Point> Generate()
```
Мне не нравится смешанная реализация перечесления и использования этого метода. Возможно лучше стоит выбрать одно из двух:
1. Сделать метод GetNextPoint(). В таком случае сам класс будет отвечать за перечисление
2. В месте вызова вызывать метод Generate() лишь один раз, и потом итерироваться уже по полученному объекту, т.е. получается что этот метод вернет тебе как бы коллекцию, которую ты и будешь перечислять.

Почему мне не нравится такая реализация:
1. вызывая метод Generate, который возвращает IEnumerable, я ожидаю что в этом IEnumerable будут содержаться все элементы без эффекта памяти.
2. Кроме того в таком случае такая реализация уязвима к многопоточному коду, да даже к однопоточному где метод Generate вызывается в разных местах. Так например ты можешь вызывать в двух местах Generate и перечисляя параллельно в них ты будешь пропускать какие-то элементы в каждом из перечислений.

Можем подробнее обсудить это в телеге

hariton27sy commented on cs/TagsCloudVisualization/Extensions/PointExtensions.cs:
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Тоже перенести в Tests

hariton27sy commented on cs/TagsCloudVisualization/Extensions/RectangleExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+﻿using System.Collections.Generic;
```
Перенести в Tests

hariton27sy commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,72 @@
+            if (rectangleSize.Height <= 0) throw new ArgumentException("Height should be positive");
+
+            var rectangle = CreateCorrectRectangle(rectangleSize);
+            ShiftRectangleToCenter(rectangle);
```
А ты уверен что этот метод работает? Посмотри чем является Rectangle struct или class?

gODeaLoAple replied:
Ахах, действительно, я не сдвигаю ректангл) Исправлю

hariton27sy commented on cs/TagsCloudVisualization/DrawerSettings.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using System;
```
Название для этого класса не очень подходящее. Все же под таким названием подразумевались бы настройки для самого Drawer'а. Для этого класса больше бы подошло CloudGenerator или что-то с похожим смыслом, так как здесь по сути происходит. Да  и на самом деле я не совсем понимаю смысла этого класса прослойки.

Давай подробнее обсудим в телеграмме, если нужно

hariton27sy commented on cs/TagsCloudVisualization/TagsCloudDrawer.cs:
```diff
@@ -0,0 +1,76 @@
+using System.Linq;
+using TagsCloudVisualization.Interfaces;
+
+#pragma warning disable CA1416
```
Давай добавим комментарий к этой директиве, иначе непонятно зачем воткнули здесь это

hariton27sy commented on cs/TagsCloudVisualization/TagsCloudDrawer.cs:
```diff
@@ -0,0 +1,76 @@
+            var bitmap = new Bitmap(imageSize.Width, imageSize.Height);
+            var graphics = Graphics.FromImage(bitmap);
+
+            graphics.FillRectangle(Brushes.Gray, 0, 0, bitmap.Width, bitmap.Height);
```
Захардкожен цвет. Давай передавать его в конструктор. В таком случае как раз помог класс с названием TagsCloudDrawerSettings со всеми конфигами drawer'a

hariton27sy commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,73 @@
+                new TagsCloudDrawer(new GrayscaleColorGenerator(rnd)));
+        }
+
+        private static Rectangle[] GenerateRectangles(int count, CircularCloudLayouter layouter, Func<Size> sizeFactory)
```
Перенеси этот метод в текущий DrawerSettings. И вызывай уже внутри него при вызове Draw. Тогда генерация(можно назвать бизнес-логикой😄) уже будет проходить при вызове соответствующего метода, а не при инициализации.

Commit:
Fix shifting rectangle to center. Now it works

Commit:
Change IVectorsGenerator interface so that now it just generate next point

Commit:
Add comments under using #pragma warning disable CA1416

Commit:
Introduce backgroundColor field to the TagsCloudDrawer ctor

Commit:
Clean up code

Commit:
Add missing comment under using #pragma warning disable

Commit:
Rename DrawerSettings to TagsCloudGenerator and put some logic there from Program.cs

Commit:
Refactor: move methods and classes that used only in Test project into there

Commit:
CleanUp some files

hariton27sy commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,72 @@
+        private Rectangle GetShiftedToCenterRectangle(Rectangle rectangle)
+        {
+            var center = rectangle.GetCenter();
+            var dir = new Point(_center.X - center.X, _center.Y - center.Y);
```
dir и shift можно сократить 😄, но не принципиально

hariton27sy commented on cs/TagsCloudVisualization/TagsCloudGenerator.cs:
```diff
@@ -0,0 +1,48 @@
+            _drawer.Draw(bitmap, GenerateRectangles(), _cloudScale);
+        }
+
+        private Rectangle[] GenerateRectangles()
```
На будущее, подумай над ленивой генерацией прямоугольников - понятно что для этого в drawer'e придется что-то поменять (возможно не пытаться вписать все облако в bitmap и пусть выходит за границы - тут уже твое дело).
Плохо это тем, что в случае большого количества прямоугольников нужно будет создавать массив большого объема - выделять память. При ленивой ты просто из метода бы доставал прямоугольник друг за другом и ничего бы не выделял.

Еще в чем минус конкретно здесь, это использование ToArray. Этот метод коварен тем, что во время исполнения постоянно будут пересоздаваться массивы -> больше затрат на аллокацию и GC.

Это на самом деле просто оптимизации, которые больше относятся к высокопроизводительному коду, но на такие моменты тоже нужно обращать внимание, чтобы писать код еще эффективнее.

Несмотря на удобства LINQ нужно использовать его с умом и понимать, где это может помешать. Особенное внимание это конечно ToArray, ToList, OrderBy, Distinct и т.д.

Commit:
Refactor inline dir variable

Commit:
Refactor make drawer work with IEnumerable<Rectangle> instead of Array<Rectangle>.
-- But now it does not fit all rectangles into image, so maybe it was not its responsibility...

Commit:
Refactor shift rectangle method
